source/extensions/tracers/datadog/agent_http_client.cc

```diff
@@ -0,0 +1,144 @@
+  for (const auto& [request_ptr, _] : handlers_) {
+    RELEASE_ASSERT(request_ptr,
+                   "null Http::AsyncClient::Request* in handler map of Datadog::AgentHTTPClient");
+    request_ptr->cancel();
+  }
+}
+
+// datadog::tracing::HTTPClient
+
+datadog::tracing::Expected<void> AgentHTTPClient::post(const URL& url, HeadersSetter set_headers,
+                                                       std::string body,
+                                                       ResponseHandler on_response,
+                                                       ErrorHandler on_error) {
+  ENVOY_LOG(debug, "flushing traces");
+
+  auto message = std::make_unique<Http::RequestMessageImpl>();
+  Http::RequestHeaderMap& headers = message->headers();
+  headers.setReferenceMethod(Http::Headers::get().MethodValues.Post);
+  headers.setReferencePath(url.path);
+  headers.setReferenceHost(reference_host_);
+
+  RequestHeaderWriter writer{message->headers()};
+  set_headers(writer);
+
+  message->body().add(body);
+  ENVOY_LOG(debug, "submitting trace(s) to {} with payload size {}", url.path, body.size());
+
+  if (!collector_cluster_.threadLocalCluster().has_value()) {
+    ENVOY_LOG(debug, "collector cluster '{}' does not exist", cluster_);
+    stats_.reports_skipped_no_cluster_.inc();
+    return datadog::tracing::nullopt;
+  }
+
+  Http::AsyncClient::Request* request =
+      collector_cluster_.threadLocalCluster()->get().httpAsyncClient().send(
+          std::move(message), *this,
+          Http::AsyncClient::RequestOptions().setTimeout(std::chrono::milliseconds(1000)));
+  if (!request) {
+    stats_.reports_failed_.inc();
+    return datadog::tracing::Error{datadog::tracing::Error::ENVOY_HTTP_CLIENT_FAILURE,
+                                   "Failed to create request."};
+  }
+
+  handlers_.emplace(request, Handlers{std::move(on_response), std::move(on_error)});
+  return datadog::tracing::nullopt;
+}
+
+void AgentHTTPClient::drain(std::chrono::steady_clock::time_point) {}
+
+nlohmann::json AgentHTTPClient::config_json() const {
+  return nlohmann::json::object({
+      {"type", "Envoy::Extensions::Tracers::Datadog::AgentHTTPClient"},
+  });
+}
+
+// Http::AsyncClient::Callbacks
+
+void AgentHTTPClient::onSuccess(const Http::AsyncClient::Request& request,
+                                Http::ResponseMessagePtr&& response) {
+  const std::uint64_t status = Http::Utility::getResponseStatus(response->headers());
+  if (status != std::uint64_t(Http::Code::OK)) {
+    stats_.reports_dropped_.inc();
+  } else {
+    ENVOY_LOG(debug, "traces successfully submitted to datadog agent");
+    stats_.reports_sent_.inc();
+  }
+
+  auto found = handlers_.find(const_cast<Http::AsyncClient::Request*>(&request));
+  if (found == handlers_.end()) {
+    ENVOY_LOG(debug, "request at address 0x{} is not in the map of {} Handlers objects",
+              absl::StrFormat("%p", &request), handlers_.size());
+    return;
+  }
```

Comment:
Should we do a handler check first here to ensure the request is sent rather than a local failure?

Reply:
What do you mean by "local failure" and "handler check"?

In light of our discussion in the unit test, maybe you're referring to the 503 "success" scenarios where a request is not actually ever sent. If so, then I'm not sure what is best. Since we have an HTTP response status, I'm tempted to consider it a "success" for the purposes of the `Handlers` callbacks. On the other hand, being able to distinguish between "sent a request and received a 503" and "never sent a request and indicated this with a 503" might be helpful for future debugging. Ideally the two cases could be distinguished, but I'm not sure they can be.

My preference is to leave the 503 ambiguity and consider it a true response.

What do you think?

Reply:
Sounds okay to me. :)